### ISSUE TITLE
Cannot upload binary attachment.

### ISSUE THREAD

**Issue by gr2m** [pouchdb/express-pouchdb#200]
I'd love to use pouchdb-server as a drop-in replacement for CouchDB across our projects, for local developent and testing. But there seems to be an issue with uploading binary files, that I couldn't understand or workaround yet.

Here's a tiny repository with instructions to reproduce the issue:
https://github.com/gr2m/pouchdb-dev-server

More than happy to help fixing the issue, or at least creating a test form `express-pouchdb` that reproduces it, I just need some guidance on how to approach it <3

**Comment by nolanlawson** [comment]
Yup, it's a known issue. You can repro it by using the "upload file" button in Futon/Fauxton as well.

I think the issue is just that we don't accept multipart/mixed uploads (only base64). If you want to look for a place to start, I think it's [here](https://github.com/pouchdb/express-pouchdb/blob/master/lib/routes/attachments.js#L30-L32).

**Comment by nolanlawson** [comment]
Ah, you may even be able to reuse [this](https://github.com/pouchdb/express-pouchdb/blob/0bdc5e07bfb4d00bf896c6ca253a929594324661/lib/routes/documents.js). It looks like we're using `multiparty` here.

**Comment by gr2m** [comment]
The code here https://github.com/pouchdb/express-pouchdb/blob/0bdc5e07bfb4d00bf896c6ca253a929594324661/lib/routes/documents.js is identical with the current code.

A problem might be that the upload is handled by https://github.com/pouchdb/express-pouchdb/blob/master/lib/routes/documents.js#L14, instead of a route in https://github.com/pouchdb/express-pouchdb/blob/master/lib/routes/attachments.js. 

Just thinking out load, hope it helps to debug, I won't have much time myself the coming 4 weeks :(

**Comment by gr2m** [comment]
I've a hacky workaround for https://github.com/pouchdb/express-pouchdb/blob/master/lib/routes/documents.js#L36:

```js
try {
  doc = JSON.parse(field);
} catch(error) {
  var attachmentName = Object.keys(doc._attachments)[0];
  doc._attachments[attachmentName].data = new Buffer(field).toString('base64');
}
```

But the uploaded file is corrupted it seems. I'm sure I'm doing something wrong with the `new Buffer(field).toString('base64')` part, `field` seems to be the binary data of the image. Any idea?

**Comment by gr2m** [comment]
If I manually convert the PNG to a string (e.g. on http://www.base64-image.de/) and paste the string directly into my code (`doc._attachments[attachmentName].data = '...base64stringhere...'`), it works.

**Comment by nolanlawson** [comment]
hmmm, maybe you need to do `new Buffer(field, 'binary').toString('base64')` ?

Also that dog describes me most of the time, lol.

**Comment by gr2m** [comment]
@nolanlawson yeah I tried that, it should work, but doesn't. I've updated my code example. I've also tried to fix it with an update to latest `node-multiparty` as there is a new major version, but no luck. Maybe it's also nano doing something wrong. `multiparty` does not emit the `file` event because `partFilename` is not set in the `multipart/related` request. I've really no idea :)

**Comment by gr2m** [comment]
If I just console log the binary string, it starts with `�PNG

IHDR�>Bf�QiCCPICC ProfileX	�Yw8��ۿ��~���`

If I do `fs.readFileSync('test.png', {encoding: 'binary'})`, the returned starts looks like `PNG\r\n\u001a\n\u0000\u0000\u0000\rIHDR\u0000\u0000\u0000ç\u0000\u0000\u0000E\b\u0002\u0000\u0000\u0000>Bf\u0000\u0000\u0018QiCCPICC Profile\u0000`

Maybe this rings a bell for someone :tm: ?

**Comment by nolanlawson** [comment]
If you console log `JSON.stringify(binaryString)`, what do you see? It may have been coerced to UTF-16/UTF-8 due to content-type headers or something.

**Comment by gr2m** [comment]
`JSON.stringify(binaryString)` statrs with `�PNG\r\n\u001a\n\u0000\u0000\u0000\rIHDR\u0000\u0000\u0000�\u0000\u0000\u0000E\b\u0002\u0000\u0000\u0000>Bf�\u0000\u0000\u0018QiCCPICC Profile\u0000\u0000X` Close, but still looking like some encoding foo.

I had to give up on it for now and use inline attachments upload with `base64` encoding as described in the [PouchDB guides](http://pouchdb.com/guides/attachments.html), that works fine

**Comment by jstjoe** [comment]
I believe this is because Node's fs is giving you a buffer object, not a binary string. Use the toString() method with a base64 encoding option if you want to store it as base64. [buffer.toString()](https://nodejs.org/api/buffer.html#buffer_buf_tostring_encoding_start_end)

**Comment by gr2m** [comment]
@jstjoe yeah I thought that, too. I tried calling `.toString('base64')` on it directly. But it seems to be somewhat corrupted. 

I think it would be good to have another test case though, that is not using `nano`. Just to make sure it's not the root of the problem here. I'm unfortunately traveling the next weeks, I wish I could help better.

Just a question though: isn't there a CouchDB test suite we could use to test this case? /cc @janl

**Comment by jstjoe** [comment]
@gr2m were you still passing {encoding: 'binary'} when you tested? That would give you a binary string back but has all kinds of issues and support is being removed.

If you remove that it *should* give you a proper Buffer object.

**Comment by gr2m** [comment]
yes, I tried it all in all combinations :)

**Comment by nolanlawson** [comment]
You can run the CouchDB test suite by doing `npm run test-couchdb`. I believe it should run all the tests, even the ones that are currently failing.

**Comment by marten-de-vries** [comment]
It actually blacklists the failing tests so it can be used to test for regressions. To enable CouchDB tests, change node_modules/couchdb-harness/index.js

**Comment by nolanlawson** [comment]
fixed by 337a8ecd2c6c5b79f9e2082ddb0c86f2bac44d59
